atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/cache/ValidatingTransactionScopedCacheTest.java

```diff
@@ -129,6 +138,60 @@ public void differentByteArrayReferencesDoNotCauseValidationToFail() {
         assertThatCode(() -> validatingCache.get(TABLE, CELLS, valueLoader)).doesNotThrowAnyException();
     }
 
+    @Test
+    public void getRowsReadsRemotelyOnceWhenValidating() {
+        ValueCacheSnapshot snapshot = ValueCacheSnapshotImpl.of(
+                HashMap.of(CellReference.of(TABLE, createCell(1, 1)), CacheEntry.unlocked(VALUE_1)),
+                HashSet.of(TABLE),
+                ImmutableSet.of(TABLE));
+        TransactionScopedCache delegate = TransactionScopedCacheImpl.create(snapshot, metrics);
+        TransactionScopedCache validatingCache = new ValidatingTransactionScopedCache(delegate, 1.0, () -> {});
+
+        ImmutableList<byte[]> rowsAndCols = ImmutableList.of(createBytes(1), createBytes(2));
+        ColumnSelection columns = ColumnSelection.create(rowsAndCols);
+
+        Function<Set<Cell>, Map<Cell, byte[]>> cellLoader = mock(Function.class);
+        Function<Iterable<byte[]>, NavigableMap<byte[], RowResult<byte[]>>> rowLoader = mock(Function.class);
+
+        // set up mock
+        NavigableMap<byte[], RowResult<byte[]>> resultMap = new TreeMap<>(UnsignedBytes.lexicographicalComparator());
+        SortedMap<byte[], byte[]> rowResults = new TreeMap<>(UnsignedBytes.lexicographicalComparator());
+        rowResults.put(createBytes(1), VALUE_1.value().get());
+        rowResults.put(createBytes(2), createBytes(2));
+        resultMap.put(createBytes(1), RowResult.create(createBytes(1), rowResults));
+        when(rowLoader.apply(any())).thenReturn(resultMap);
+
+        validatingCache.getRows(TABLE, rowsAndCols, columns, cellLoader, rowLoader);
+
+        verifyNoInteractions(cellLoader);
+        verify(rowLoader, times(1)).apply(any());
+    }
+
+    @Test
+    public void getRowsValidates() {
+        ValueCacheSnapshot snapshot = ValueCacheSnapshotImpl.of(
+                HashMap.of(CellReference.of(TABLE, createCell(1, 1)), CacheEntry.unlocked(VALUE_1)),
+                HashSet.of(TABLE),
+                ImmutableSet.of(TABLE));
+        TransactionScopedCache delegate = TransactionScopedCacheImpl.create(snapshot, metrics);
+        TransactionScopedCache validatingCache = new ValidatingTransactionScopedCache(delegate, 1.0, () -> {});
+
+        ImmutableList<byte[]> rowsAndCols = ImmutableList.of(createBytes(1));
+        ColumnSelection columns = ColumnSelection.create(rowsAndCols);
+
+        Function<Set<Cell>, Map<Cell, byte[]>> cellLoader = mock(Function.class);
+        Function<Iterable<byte[]>, NavigableMap<byte[], RowResult<byte[]>>> rowLoader = mock(Function.class);
+
+        when(rowLoader.apply(any())).thenReturn(new TreeMap<>(UnsignedBytes.lexicographicalComparator()));
+
+        assertThatThrownBy(() -> validatingCache.getRows(TABLE, rowsAndCols, columns, cellLoader, rowLoader))
+                .isExactlyInstanceOf(TransactionLockWatchFailedException.class)
+                .hasMessage("Failed lock watch cache validation - will retry without caching");
```

Comment:
It'd be nice to have a test that tests the other error message (where the rows mismatch, as opposed to the cells).

Reply:
It's the same message